### PR TITLE
generate_binary: Add NX COMPAT flag manually when genpeimg missing

### PR DESCRIPTION
**Pull request by nfrayer**
When genpeimg or python3-pefile is missing, add the NX COMPAT flag manually by setting bit4 of the DllCharacteristics in the optional header.

Signed-off-by: Nicolas Frayer <nfrayer@redhat.com>

### CODE REVIEW

File: efi/generate_binary.py [hunk 1]
```diff
@@ -73,8 +73,19 @@ def _add_nx_pefile(args):
     try:
         import pefile
     except ImportError:
-        print("Unable to add NX support to binaries without genpeimg or python3-pefile")
-        sys.exit(1)
+        print("Unable to add NX support to binaries without genpeimg or python3-pefile, adding NX support manually")
+        import os
```
Comment (hughsie) [inline efi/generate_binary.py:77]: i think this should be at the top with the other imports

File: efi/generate_binary.py [hunk 2]
```diff
@@ -73,8 +73,19 @@ def _add_nx_pefile(args):
...
+        with open(args.outfile, 'r+b') as fh:
```
Comment (hughsie) [inline efi/generate_binary.py:78]: for my tiny brain, what's `r+b`? I did expect to see a `rwb` or something.

Comment (nfrayer) [reply efi/generate_binary.py:78]: ```r+b``` is for read and write in binary mode. ```+``` allows writing as well as far as I know.

File: efi/generate_binary.py (hunk 1 above)
Comment (nfrayer) [reply efi/generate_binary.py:77]: Ok, I'll change that.

**Comment by superm1** [comment]
> @superm1 the context here is that we don't have genpeimg or python3-pefile in the RHEL 9 buildroot.

It's not possible to add them to the spec file?

File: efi/generate_binary.py [hunk 3]
```diff
@@ -73,8 +73,19 @@ def _add_nx_pefile(args):
     try:
         import pefile
     except ImportError:
-        print("Unable to add NX support to binaries without genpeimg or python3-pefile")
-        sys.exit(1)
+        print("Unable to add NX support to binaries without genpeimg or python3-pefile, adding NX support manually")
```
Comment (superm1) [inline efi/generate_binary.py:76]: It's no longer true that it can't be done. You're doing it!

File: efi/generate_binary.py [hunk 4]
```diff
@@ -73,8 +73,19 @@ def _add_nx_pefile(args):
...
+        import os
+        with open(args.outfile, 'r+b') as fh:
+                buf = bytearray(fh.read(os.path.getsize(args.outfile)))
+                fh.seek(0)
+                # Set bit 4 (IMAGE_DLLCHARACTERISTICS_NX_COMPAT) of DllCharacteristics within the COFF optional header
+                # DllCharacteristics offset = 0xde size = 2 bytes
+                # https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#optional-header-image-only
+                buf[0xdf] |= 0x1
+                fh.write(buf)
+                fh.close()
```
Comment (superm1) [inline efi/generate_binary.py:86]: This isn't really needed when you're doing `with` to open is it?

Comment (hughsie) [reply efi/generate_binary.py:86]: Correct, close isn't needed.

**Comment by hughsie** [comment]
> It's not possible to add them to the spec file?

Nah, they're literally unavailable unfortunately. One to fix for RHEL 10.

File: efi/generate_binary.py (hunk 2 above)
Comment (hughsie) [reply efi/generate_binary.py:78]: TIL, thanks!

**Comment by hughsie** [comment]
@nfrayer also, CI doesn't like the change -- can you run `black` on the source file please.

**Comment by superm1** [comment]
> > It's not possible to add them to the spec file?
> 
> Nah, they're literally unavailable unfortunately. One to fix for RHEL 10.

Well if we're switching the build system to manually do this I don't think it ever gets fixed :(
But it is what it is.

**Comment by hughsie** [comment]
> Well if we're switching the build system to manually do this I don't think it ever gets fixed

Less pessimism! :) I've already added the python-pefile dep to Fedora *mumble*, and that's what'll be forked into RHEL 10.

**Comment by superm1** [comment]
> > Well if we're switching the build system to manually do this I don't think it ever gets fixed
> 
> Less pessimism! :) I've already added the python-pefile dep to Fedora _mumble_, and that's what'll be forked into RHEL 10.

In that case can this build system change just be in a RHEL9 branch perhaps?

**Comment by hughsie** [comment]
> In that case can this build system change just be in a RHEL9 branch perhaps?

I'd be 100% fine with that. Lets review it here, get it 100% then put it in something like a `rhel9` branch?

**Comment by superm1** [comment]
> > In that case can this build system change just be in a RHEL9 branch perhaps?
> 
> I'd be 100% fine with that. Lets review it here, get it 100% then put it in something like a `rhel9` branch?

sounds great, thanks.

File: efi/generate_binary.py (hunk 4 above)
Comment (nfrayer) [reply efi/generate_binary.py:86]: Ok I wan't sure, will remove it.

File: efi/generate_binary.py (hunk 3 above)
Comment (nfrayer) [reply efi/generate_binary.py:76]: I'll rephrase it :)

**Comment by nfrayer** [comment]
> @nfrayer also, CI doesn't like the change -- can you run `black` on the source file please.

Yes I just got black, will run it before my next commit.

File: efi/generate_binary.py [hunk 5]
```diff
@@ -66,15 +73,75 @@ def _run_genpeimg(args):
...
+        with open(args.outfile, "r+b") as fh:
+            buf = bytearray(fh.read(os.path.getsize(args.outfile)))
+            fh.seek(0)
+            # Set bit 4 (IMAGE_DLLCHARACTERISTICS_NX_COMPAT) of DllCharacteristics within the COFF optional header
+            # DllCharacteristics offset = 0xde size = 2 bytes
+            # https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#optional-header-image-only
+            DllCharacteristics = (
+                buf[OPTIONALHDR_DLLCHARACTERISTICS + 1] << 8
+                | buf[OPTIONALHDR_DLLCHARACTERISTICS]
+            )
```
Comment (hughsie) [inline efi/generate_binary.py:131]: could you use struct.unpack_from perhaps?

Comment (nfrayer) [reply efi/generate_binary.py:131]: ah thanks, it's the function I was looking for to replace this ...

**Comment by hughsie** [comment]
I've pushed a tweaked version of this to the `rhel9` branch. I'll close this one now, but thanks @nfrayer !

**Comment by nfrayer** [comment]
Thanks @hughsie @superm1 for the reviews !

**Pull request closed without merging**